teamscale_client/client.py

```diff
@@ -108,157 +142,205 @@ def put(self, url, json=None, parameters=None, data=None):
 
         Returns:
             requests.Response: request's response
 
         Raises:
             ServiceError: If anything goes wrong
         """
-        response = requests.delete(url, params=parameters, auth=self.auth_header, verify=self.sslverify,
-                                   timeout=self.timeout)
+        response = requests.delete(url, auth=self.auth_header, verify=self.sslverify, timeout=self.timeout, **kwargs)
         if not response.ok:
-            raise ServiceError("ERROR: PUT {url}: {r.status_code}:{r.text}".format(url=url, r=response))
+            raise ServiceError(f"ERROR: DELETE {url}: {response.status_code}:{response.text}")
         return response
 
-    def add_findings_group(self, name, mapping_pattern):
+    def add_findings_group(self, name: str, mapping_pattern: str) -> requests.Response:
         """Adds group of findings.
 
         Args:
-            name (str): Name of group.
-            mapping_pattern (str): Regular expression to match a finding's ``typeid`` in order to belong to this group.
+            name: Name of group.
+            mapping_pattern: Regular expression to match a finding's ``typeid`` in order to belong to this group.
         Returns:
             requests.Response: request's response
         """
-        url = "%s/%s" % (self.get_global_service_url('external-findings-group'), name)
-        return self.put(url, {'groupName': name, 'mapping': mapping_pattern})
+        return self.post(
+            f"{self._api_url_version}/external-findings/groups",
+            json={'groupName': name, 'mapping': mapping_pattern}
+        )
 
-    def add_finding_descriptions(self, descriptions):
+    def add_finding_descriptions(self, descriptions: List[FindingDescription]) -> requests.Response:
         """Adds descriptions of findings.
 
         Args:
-            descriptions (list): List of :class:`FindingDescription` to add to Teamscale.
+            descriptions: List of :class:`FindingDescription` to add to Teamscale.
         Returns:
             requests.Response: request's response
         """
-        base_url = self.get_global_service_url('external-findings-description')
         response = None
         for finding_description in descriptions:
-            some_description = dict()
-            some_description['typeId'] = finding_description.typeid
-            some_description['description'] = finding_description.description
-            some_description['enablement'] = finding_description.enablement
-            some_description['name'] = finding_description.name
-            url = "%s/%s" % (base_url, finding_description.typeid)
-            response = self.put(url, some_description)
-            if response.text != 'success':
-                return response
-
+            response = self.post(
+                f"{self._api_url_version}/external-findings/descriptions",
+                data=to_json(finding_description),
```

Comment:
Sorry for the stupid question, but can't you use

```suggestion
                json=finding_description,
```

here as well like in the previous call.

Reply:
No stupid question at all, It took me some hours to realize that the parameters `data` and `json`actually expect different formats.

 - `data` takes anything and inserts it directly into the body
```python
# fine
data=to_json(finding_description)
```
 - `json` takes anything and serializes it with the default encoder. However, the default encoder only supports the base types
```python
# would raise an exception
json=finding_description 
# in this case fine, but it will fail for nested objects (e.g. FileFindings contains
# the class Finding with the former being translated to a dict, but the member not)
json=vars(finding_description)
# fine, with method to_dict from the utils folder since any (nested) object is converted to a dict
json=to_dict(finding_description)
```

However, `to_dict(x)` is basically `json.loads(to_json(x))`.

So `data=to_json(finding_description)` is the cheapest solution, which works for any objects independent how nested thy are and `json=x` only where x is a `dict` which does not contain any custom objects.